src/rust/dns/lib.rs

```diff
@@ -0,0 +1,204 @@
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DnsParserError {
+    #[error("Invalid hex string: {0}")]
+    InvalidHexString(String),
+    #[error("ParseInt error: {0}")]
+    ParseIntError(#[from] std::num::ParseIntError),
+    #[error("Invalid DNS response: {0}")]
+    InvalidDnsResponse(String),
+    #[error("unknown dns parser error")]
+    Unknown,
+}
+
+#[cxx::bridge(namespace = "workerd::rust::dns")]
+mod ffi {
+    /// CAA record representation
+    struct CaaRecord {
+        critical: u8,
+        field: String,
+        value: String,
+    }
+    /// NAPTR record representation
+    struct NaptrRecord {
+        flags: String,
+        service: String,
+        regexp: String,
+        replacement: String,
+        order: u32,
+        preference: u32,
+    }
+    extern "Rust" {
+        fn parse_caa_record(record: &str) -> Result<CaaRecord>;
+        fn parse_naptr_record(record: &str) -> Result<NaptrRecord>;
+    }
+}
+
+/// Given a vector of strings, converts each slice to UTF-8 from HEX.
+///
+/// # Errors
+/// `DnsParserError::InvalidHexString`
+/// `DnsParserError::ParseIntError`
+pub fn decode_hex(input: &[&str]) -> Result<Vec<String>, DnsParserError> {
+    let mut v = Vec::with_capacity(input.len());
+
+    for slice in input {
+        let num = u16::from_str_radix(slice, 16)?;
+        let ch = String::from_utf16(&[num])
+            .map_err(|_| DnsParserError::InvalidHexString("Invalid UTF-16 sequence".to_owned()))?;
```

Comment:
you don't need map_err if you have thiserror with `from` setup correctly.

Reply:
I didn't want to add a specific error for utf16 encoding. invalid hex string with appropriate error message already points us towards this particular line.